rfd/0144-client-tools-auto-updates.md

```diff
@@ -0,0 +1,82 @@
+
+### Version discovery
+
+Cloud will host the cloud-stable version of Teleport at https://updates.releases.teleport.dev/v1/stable/cloud/client-version. This endpoint will serve a Teleport version e.g. `v13.2.3`.
+
+The client tools are not using the same endpoint that agents use to discover the cloud-stable version of Teleport because having a separate channel allows us to rollback a botched client release without having to rollback the cluster or agents at the same time.
+
+The stable version should be updated after each Cloud Tenant upgrade. This should follow the same workflow as updating the cloud-stable agent version endpoint. The version should be updated soon after Cloud Tenant upgrades have been completed.
+
+Unlike the agent version discovery model, there will not be an endpoint to identify if a critical update is available. The reason for this is because updates are not executed in a scheduled update window. Teleport client tools will check this endpoint on every login and update accordingly if a new version is available.
+
+On every login, tsh will pull the current cloud-stable version of Teleport and store the value in the .tsh directory under `$HOME/.tsh/cloud-stable-version`.
+
+### Caching
+
+A single version of the Teleport client tools will be stored in the user's $HOME directory. Teleport already makes use of a .tsh directory for storing tsh config. Cached versions of tsh/tctl will also live within this directory under ` $HOME/.tsh/cache/<version>/{tsh,tctl}`. Whenever a new version of the client tools are available, the existing cached version will be replaced.
+
+On every login, tsh will compare the currently cached version of the client tools with the current stable version stored in `$HOME/.tsh/cloud-stable-version`. If the cached version is behind the current stable version, a new version of the client tools will be downloaded to replace the existing ones. When downloading the Teleport client tools, the binaries will be downloaded directly. We need to make sure to download a version that is compatible with the system.
+
+Because only a single version of the client tools will be cached at a time, we do not need to worry about cleaning up the cache.
+
+### Package upgrade
+
+As an alternative to caching, we've considered directly upgrading the Teleport packages.
+
+This approach will be difficult to implement because we support multiple package repositories and installation methods. The amount of maintenance and testing required to support this solution is probably not worth the effort at this time.
+
+Another downside to this approach would be that upgrades would require sudo/admin privileges.
+
+### Config
+
+By default, the client tools will have auto-updates enabled. I don't see compelling reason to allow users to disable auto updates for client tools. But if we want to allow this feature to be disabled, we can add support in the tsh configuration files by adding a DISABLE_AUTO_UPDATES flag.
```

Comment:
Right, we should have an escape hatch just in case. I wonder if we can have a hidden (and probably undocumented) flag. Then users who really want to disable it can write a tsh alias.